crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+use super::*;
+use bindings::*;
+use std::sync::*;
+
+/// A type that you can use to declare and implement an event of a specified delegate type.
+///
+/// The implementation is thread-safe and designed to avoid contention between events being
+/// raised and delegates being added or removed.
+pub struct Event<T: Interface + Clone> {
+    swap: Mutex<()>,
+    change: Mutex<()>,
+    delegates: Array<T>,
+}
+
+impl<T: Interface + Clone> Default for Event<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Interface + Clone> Event<T> {
+    /// Creates a new, empty `Event<T>`.
+    pub fn new() -> Self {
+        Self { delegates: Array::new(), swap: Mutex::default(), change: Mutex::default() }
+    }
+    /// Registers a delegate with the event object.
+    pub fn add(&mut self, delegate: &T) -> Result<i64> {
+        let mut _lock_free_drop = Array::new();
+        Ok({
+            let change_lock = self.change.lock().unwrap();
+            let mut new_delegates = Array::with_capacity(self.delegates.len() + 1)?;
+            for delegate in self.delegates.as_slice() {
+                new_delegates.push(delegate.clone());
```

Comment:
Why do the delegates have to be cloned instead of the new delegate just being pushed on to the existing delegates array?

Reply:
To avoid disrupting a concurrent read of the delegates, a new array needs to be created every time. The tension is between a thread adding/removing a delegate and another thread firing the event and thus cycling through the existing array. The vast majority of event sources have either zero or one handler, and this was determined to be the most efficient implementation.